libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -31,24 +32,20 @@ public class MicrosoftAppCredentials : ServiceClientCredentials
         /// The key for Microsoft app Password.
         /// </summary>
         public const string MicrosoftAppPasswordKey = "MicrosoftAppPassword";
+        
+        private static HttpClient _httpClient = new HttpClient();
 
-        /// <summary>
-        /// The token refresh code uses this client. Ideally, this would be passed in or set via a DI system to 
-        /// allow developer control over behavior / headers / timeouts and such. Unfortunately this is buried
-        /// pretty deep, the static solution used here is much cleaner. If this becomes an issue we could
-        /// consider circling back and exposing developer control over this HttpClient. 
-        /// </summary>
-        private static readonly HttpClient _httpClient = new HttpClient();
-
-        private static object _trustedHostNamesSync = new object();
-        private static readonly IDictionary<string, DateTime> _trustedHostNames = new Dictionary<string, DateTime>()
-        {
-            { "state.botframework.com", DateTime.MaxValue },
-            { "api.botframework.com", DateTime.MaxValue}
-        };
+        protected static ConcurrentDictionary<string, DateTime> TrustedHostNames = new ConcurrentDictionary<string, DateTime>(
+                                                                                        new Dictionary<string, DateTime>() {
+                                                                                            { "state.botframework.com", DateTime.MaxValue },
+                                                                                            { "api.botframework.com", DateTime.MaxValue },
+                                                                                            { "token.botframework.com", DateTime.MaxValue }
+                                                                                        });
 
-        private static object _cacheSync = new object();
-        protected static readonly IDictionary<string, OAuthResponse> _cache = new Dictionary<string, OAuthResponse>();
+        protected static readonly Dictionary<string, Task<OAuthResponse>> tokenTaskCache = new Dictionary<string, Task<OAuthResponse>>();
```

Comment:
All of the concurrent dictionaries that I removed in the last code pass, as they were largely not needed and I moved to a simpler concurrency model, seem to have been lost. 

We should consider keeping the previous code, and not bringing back the *much* older v3 code that's in this merge.

Reply:
this is all new code that I spent that month debugging, I really really don't want to change this.

---
In reply to: [198676497](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198676497) [](ancestors = 198676497)